### DISCUSSION TITLE
WindTerm 2.2.0

### DISCUSSION THREAD

**Discussion by kingToolbox** [kingToolbox/WindTerm#272] [General]
# Usage:
### Windows
Please install **[vc_redist.x86](https://aka.ms/vs/16/release/vc_redist.x86.exe)** if WindTerm **failed** to run.
Please use **WindTerm_Win10_Conpty.exe** to support Windows 10 shell. (**Windows 19H1 [v10.0.18309] and later**)

### MacOS
Please open dmg file and drag WindTerm.app to the `Applications` folder.

### Linux
Please use the following command to run the WindTerm after downloading.
``` shell
tar -xzvf WindTerm_2.2.0_Linux_Portable.tar.gz && cd ./WindTerm_2.2.0 && chmod +x ./WindTerm && ./WindTerm
```

### Upgrade
For Windows and Linux, copy or move the **profiles** folder from the **old Windows/Linux version** to the **new Windows/Linux version** and all the sessions and configuration will be restored.

For MacOS, copy or move the **profiles** folder from the old **WindTerm.app/Contents/MacOS** to the new **WindTerm.app/Contents/MacOS** and all the sessions and configuration will be restored.

The **profiles** folder is cross-platform. 

# Note:

**Sorry, since the MacOS version has not been stable so far, the release of the WindTerm 2.2.0 MacOS version has been cancelled and it is expected that a brand new stable `universal app` will be released in December.**

# Known Issues:
- [ZMODEM] **Only support file names contains ASCII characters in Windows, otherwise the file name of the transfered file will be garbled after transmission, but it does not affect the correctness of the file content.** Linux and MacOS version does not have this problem.

# Translation:
- [CHINESE] Update the language file, contributed by **MosamXu@github**. #246 

# Features:
- [GUI] **Supports Multilingual User Interface.** #6 #22 #43 #88 #114 #168 #176 #211 #217 
    - Arabic - عربي
    - Bengali - বাংলা
    - Chinese (Simplified) - 中文 (简体)
    - Chinese (Traditional) - 正體中文 (繁體)
    - Czech - Čeština
    - Danish - Dansk
    - Dutch - Nederlands
    - English - English
    - Finnish - suomi
    - French - Français
    - German - Deutsch
    - Greek - Ελληνικά
    - Hebrew - עברית
    - Hindi - हिन्दी
    - Hungarian - magyar
    - Icelandic - íslenska
    - Indonesian - Bahasa Indonesia
    - Italian - Italiano
    - Japanese - 日本語
    - Khmer - ខ្មែរ
    - Korean - 한국어
    - Malay - Melayu
    - Norwegian (Bokmål) - Norsk bokmål
    - Persian - فارسی
    - Polish - Polski
    - Portuguese - Português
    - Punjabi (India) - ਪੰਜਾਬੀ (ਭਾਰਤ)
    - Russian - Русский
    - Spanish - Español
    - Swedish - Svenska
    - Thai - ไทย
    - Turkish - Türkçe
    - Ukrainian - Українська
    - Urdu - اُردو
    - Vietnamese - Tiếng Việt
- [GUI] **Added the global setting dialog, `Menubar - Session (1) - Preferences - Settings`.**
- [GUI] **Supports setting the default font of the text view, `Settings - Text - Font`.** #148 
- [GUI] **Supports nesting session groups when the group name contains `>`, like `x>y>z`.** #91 #177 
- [GUI] **Supports opening a specific session or set of sessions on startup.** #108 
- [GUI] The address bar and the menu of `Add Tab` button supports nesting session groups. 
- [GUI] Supports alternate background colors for text lines, `Settings - Text - Alternate line background color`. 
- [GUI] Supports setting whether to show the `Paste` dialog on the main screen, `Settings - Terminal - Show paste dialog on the main screen`. #214
- [GUI] Added an identity file path input box in the `Session Setting - SSH - Authentication` dialog. 
- [TAB] Added `Duplicate Tab` to the customizable double-click operations of the tab. #258
- [SSH] **Supports SSH ControlMaster.** #173 
- [SSH] Supports setting mouse tracking events. #144 
- [SHELL] **Added a new `Shell` pane makes it more easier to use shells, such as Bash, Cmd, Powershell, WSL, etc.**
- [TERM] **Added `GB2312`, `GBK` and many Single-Byte Character Sets.** #208 #264 
    - GB2312
    - GBK
    - hp-roman8
    - IBM850
    - IBM866
    - IBM874
    - KOI8-U
    - macintosh
    - TSCII
    - TIS-620
    - Windows-1250
    - Windows-1251
    - Windows-1252
    - Windows-1253
    - Windows-1254
    - Windows-1255
    - Windows-1256
    - Windows-1257
    - Windows-1258
    - WINSAMI2
- [TERM] Supports `Xterm formatOtherKey` to allow the format of the escape sequence used to report modifieed keys to be overrided.
- [TERM] Supports `OSC 12` to set the text cursor color.
- [TERM] Supports `OSC 112` to reset the text cursor color. #160 #188 #229 
- [TERM] Supports `OSC 104` to reset the palette colors.
- [TERM] Supports `CSI > m` to reset key modifier options. #232 
- [TERM] Supports `CSI >  1 ; Ⓝ m` to select format for modifier keys in cursor key press reports. #232 
- [TERM] Supports `CSI >  2 ; Ⓝ m` to select format for modifier keys in function-key key press reports. #232 
- [TERM] Supports `CSI >  4 ; Ⓝ m` to select format for modifier keys in other key press reports. #232 
- [TERM] Supports setting the number of lines each time the mouse wheel scrolls in the alternative screen. #183 
- [TERM] **Supports `hard reset`, `soft reset`, and `character set reset` of the term.** #162 #199 
- [QUICKBAR] **Added a shortcut menu containing `Modify` and `Remove` actions to the command button to quickly manage the button.** 
- [MESSAGEBAR] Supports simultaneous display of multiple messages in expanded mode. 

# Improvements:
- [APP] **Redesigned, rewritten, and refactored more than 15,000 lines of core code to enhance flexibility, reliability and robustness.** 
- [GUI] **Improve the column selection and copy in auto-wrap mode.** #250 
- [GUI] Supports popping up the right-click menu in the blank space of the `Explorer Pane`.
- [GUI] Disable the pop-up context menu of the term view in the login wizard.
- [GUI] Restore the last configuration when the `Log Dialog` is opened. #240 
- [GUI] The font combo box in the settings dialog uses bold type to indicate fixed-width fonts.
- [GUI] Change the `Yes To All` button of the Close Tab dialog box to the `Don't show again` check box. 
- [GUI] Remove the extra 10% of the line height to display more lines on the screen and make the table lines and so on look more continuous. #146 
- [GUI] Remove unnecessary scheme style `terminal.text`, `terminal.text.1` and `terminal.text.2` to avoid interference with the default font settings. #202
- [GUI] Supports restoring the `Focus Mode` after restarting. #175 #197 
- [GUI] Supports restoring the visibility of the menu bar and status bar after restarting. #174 
- [FILER] Restore the last dialog geometry when opening the `upload` and `download` dialog. #247 
- [FILER] When downloading a single file, the menu item `Open Target Directory` will select the file when opening the target folder. 
- [LOG] The log file name template supports including IP addresses. #227 
- [SESSION] Supports pressing `Enter` to reconnect the session when the error "Remote channel is closed" is detected.
- [TERM] **Add and modify hundreds of pre-defined key values under `Normal keypad mode` and `Application keypad mode` of `vt52`, `vt100`, `vt102`, `vt220`, `vt340`, `vt420`, `520`, `xterm` and `xterm-256color`.** #171 
- [MESSAGEBAR] Added the time stamp of the message to the tooltip. 

# Bugs:
- [APP] **The application may crash when closing a connecting session.** #206 
- [GUI] **Sometimes the text is automatically selected by mistake when moving the mouse.** #138 
- [GUI] After switching to hexadecimal mode and switching back, the caret becomes a vertical bar.
- [GUI] When dragging the vertical scroll bar to the starting position of the column selection of hundres of lines, the carets of some lines does not blink correctly. 
- [GUI] Double-clicking the folding margin cannot correctly select the text in the folding area. 
- [GUI] When the average width of the font is less than the width of the space, it may cause unnecessary text wrapping. #171 
- [GUI] The shortcut key `Alt+A` cannot be used to check/uncheck the `Auto login next time` check box in the login wizard. #241
- [GUI] When the session name is the same as the group name, the session node is not visible in `Session Pane`. #225 
- [GUI] The title of the rename group dialog should be `Rename Group` instead of `Add Group`. #191 
- [GUI] After restarting in the focus mode, `QuickBar` does not display the groups correctly. #204 
- [GUI] The password input box in the lock screen cannot get the focus when the window is activated. #218 
- [GUI] Cannot hide the `Symbol` margin. #221 
- [GUI] When the `Paste` dialog was initially opened, the number of characters and lines of the pasted text could not be displayed correctly. 
- [GUI] When the browse buttons of `X/Y/Z Modem - Download/Upload Directory` in the `Session Settings` dialog are clicked, the browse dialog is not displayed. 
- [GUI] When all sessions are closed, the `Local/Remote` button in the status bar is not hidden correctly. 
- [GUI] Executing `Close All Views` will cause a small amount of memory leaks. 
- [LINUX] **Cannot use `fcitx` input method to input Chinese.** #39 #125 #203 
- [LINUX] Cannot minimize the window correctly after turning off always on top mode and focus mode. #198 
- [LOG] The duplicated session cannot create the log file correctly.
- [LOG] When exiting or disconnecting a session, the log file sometimes fails to record all the output. 
- [SHELL] When no Windows subsystem is installed, a large number of invalid `Linux Distributions` will be imported. #243 
- [SHELL] Incorrect `wsl -l` output parsing may result in incomplete import of all installed `Linux Distributions`.
- [SSH] Cannot log in using an identity file in a path that contains non-English characters. #159 
- [TERM] Executing `set nowrap` in `vim` will cause the first characters of wrapped lines to be displayed incorrectly. #260 
- [TERM] When automatically wrapping the long lines, the actual displayed characters in each line is one more than the number of columns of the window. 
- [TERM] `CSI J`, `CSI 0 J` and `CSI 1 J` do not place the text erased from the top line in the scroll history. #124 
- [TERM] Wrong `left arrow`, `up arrow`, `right arrow` and `down arrow` key values make it impossible to use the arrow keys in `Midnight Commander`. #171 
- [TERM] The length of the characters at the end of the `GB18030` encoded string is calculated incorrectly and causes garbled characters. #264 
- [FILER] When the language is set to `Chinese`, the file type in the `Property` dialog is displayed as garbled characters. #242 
- [EXPLORER] The synchronization channel marks that were enabled during hiding could not be displayed correctly.
- [EXPLORER] Sometimes the icon and text of the list item are not displayed completely. #256
- [EXPLORER] When the vertical scroll bar is visible, the close button is not displayed completely. 
- [QUICKBAR] When the `Rename Group` dialog box is canceled, the group name will be cleared. #166 #191 
- [QUICKBAR] When multiline text is sent to Windows Cmd, the carriage return of each line is not executed correctly. 
- [QUICKBAR] When no group is set to be visible, the `Quick` button cannot pop up the menu correctly. 

# Roadmap of v2.3 (Mid-December, for reference only)
- Automatic Completion
- Command Snippet [Description](https://github.com/kingToolbox/WindTerm/issues/239#issuecomment-951934488)
- Move Caret By Mouse #239
- Free Type Mode
- SSH Agent Forwaring (Postponed to a later version )
- New memory allocator and manager and garbage collector. (Postponed to a later version )
- Search in sessions. (Postponed to a later version )

<hr /><em>This discussion was created from the release <a href='https://github.com/kingToolbox/WindTerm/releases/tag/2.2'>WindTerm 2.2.0</a>.</em>

**Comment by kingToolbox** [comment]
Thanks to everyone who cares and helps, without you, there would be no new `WindTerm_2.2.0` version released. :heart:

**Comment by eugenov** [comment]
Hello, thanks for hard work, I believe WT will be universal terminal tool! I wait patiently for macros :)

**Reply by kingToolbox** [reply]
Thank you, you are welcome. If all goes well, the MacOS universal app will make progress next week, and I hope to release it as soon as possible. :smile:

**Reply by kingToolbox** [reply]
Sorry, macros, not MacOS.

From January next year, I will focus on the development of semi-automatic functions, including `marcos`, `triggers`, `events`, `notifications`, `external tools` and so on. I believe I won't let you down!

**Reply by xywf221** [reply]
hi, do you have the latest development of macos.I've been waiting for a long time

**Reply by kingToolbox** [reply]
Please wait patiently for a while, I should be able to start developing `macros` feature early next year. Please stay tuned, thank you.

**Reply by xywf221** [reply]
Since I switched from windows to MacOS, I haven't encountered any easy-to-use SSH tools. This is an amazing project. Please insist

**Reply by kingToolbox** [reply]
Sorry, I made a mistake again. This time it's `macos` instead of `macros`. :sweat_smile:  The `MacOS` version  has already started to be compiled, and if everything goes well, it is expected to be released as soon as mid-December. I hope `WindTerm` can bring a little convenience to your work.

**Reply by xywf221** [reply]
😄

**Reply by kingToolbox** [reply]
Hi, I am glad the new MacOS version [WindTerm_2.3.0_Prerelease_2_Mac_Portable](https://github.com/kingToolbox/WindTerm/releases/2.3-prerelease) was released, you can download and check it, thank you.

**Comment by eugenov** [comment]
Mean no offence, but this repo is spammed with issues in Chinese. Guys, English is international IT language, I cannot work with them as I need to translate every chinese issue to search for resolving or if I need to post mine (to not duplicate). Let's respect each other. Peace
